sources/lib/api/gcp/data/__init__.py

```diff
@@ -27,10 +27,109 @@ def sql(sql_template, **kwargs):
     sql_template: the template of the SQL statement with named placeholders.
     **kwargs: the dictionary of name/value pairs to use for placeholder values.
   Returns:
-    The formatted SQL statement with placeholders replaced with their values, and
-    an array of Javascript UDFs referenced by the code.
+    The formatted SQL statement with placeholders replaced with their values.
   Raises:
     Exception if a placeholder was found in the SQL statement, but did not have
     a corresponding argument value.
   """
   return _util.Sql.format(sql_template, kwargs)
+
+
+def tokenize(sql):
+  """ This is a crude tokenizer for our limited purposes.
```

Comment:
I have an alternative implementation; do you prefer this:

def _next_token(sql):
      i = 0
    
      # We use some lambda's to make the logic more clear. The start_* functions return
      # true if i is the index of the start of that construct, while the end_* functions
      # return true if i point to the first character beyond that construct or the end of the
      # content.
      #
      # We don't currently need numbers so the tokenizer here just does sequences of
      # digits as a convenience to shrink the total number of tokens. If we needed numbers
      # later we would need a special handler for these much like strings.
    
      start_multi_line_comment = lambda s, i: s[i] == '/' and i < len(s) - 1 and s[i + 1] == '*'
      end_multi_line_comment = lambda s, i: s[i - 2] == '*' and s[i - 1] == '/'
      start_single_line_comment = lambda s, i: s[i] == '-' and i < len(s) - 1 and s[i + 1] == '-'
      end_single_line_comment = lambda s, i: s[i - 1] == '\n'
      start_whitespace = lambda s, i: s[i].isspace()
      end_whitespace = lambda s, i: not s[i].isspace()
      start_number = lambda s, i: s[i].isdigit()
      end_number = lambda s, i: not s[i].isdigit()
      start_identifier = lambda s, i: s[i].isalpha() or s[i] == '_' or s[i] == '$'
      end_identifier = lambda s, i: not(s[i].isalnum() or s[i] == '_')
      start_string = lambda s, i: s[i] =='"' or s[i] == "'"
      always_true = lambda s, i: True
    
      while i < len(sql):
        start = i
        if start_multi_line_comment(sql, i):
          i += 2
          end_checker = end_multi_line_comment
        elif start_single_line_comment(sql, i):
          i += 2
          end_checker = end_single_line_comment
        elif start_whitespace(sql, i):
          i += 1
          end_checker = end_whitespace
        elif start_identifier(sql, i):
          i += 1
          end_checker = end_identifier
        elif start_number(sql, i):
          i += 1
          end_checker = end_number
        elif start_string(sql, i):
          # Special handling here as we need to check for escaped closing quotes.
          quote = sql[i]
          end_checker = always_true
          i += 1
          while i < len(sql) and sql[i] != quote:
            i += 2 if sql[i] == '\\' else 1
          i += 1  # skip quote
        else:
          # We return single characters for everything else
          i += 1
          end_checker = always_true
    
        while i < len(sql) and not end_checker(sql, i):
          i += 1
    
        (yield sql[start:i])